src/executor/stack/mod.rs

```diff
@@ -88,59 +167,71 @@ pub struct PrecompileOutput {
 	pub logs: Vec<Log>,
 }
 
-/// Precompiles function signature. Expected input arguments are:
-///  * Address
-///  * Input
-///  * Context
-///  * State
-///  * Is static
-type PrecompileFn<S> = fn(
-	H160,
-	&[u8],
-	Option<u64>,
-	&Context,
-	&mut S,
-	bool,
-) -> Option<Result<PrecompileOutput, ExitError>>;
+/// Precompiles trait which allows for the execution of a precompile.
+pub trait Precompiles {
+	/// Runs the precompile at the given address with input and context.
+	fn run<S>(
+		&self,
+		address: H160,
+		input: &[u8],
+		gas_limit: Option<u64>,
+		context: &Context,
+		state: &mut S,
+		is_static: bool,
+	) -> Option<Result<PrecompileOutput, ExitError>>;
+
+	/// Returns the set of precompile addresses.
+	fn addresses(&self) -> &[H160];
+}
+
+#[derive(Default)]
+pub struct NoPrecompile {
+	addresses: Vec<H160>,
+}
+
+impl Precompiles for NoPrecompile {
+	fn run<S>(
+		&self,
+		_address: H160,
+		_input: &[u8],
+		_gas_limit: Option<u64>,
+		_context: &Context,
+		_state: &mut S,
+		_is_static: bool,
+	) -> Option<Result<PrecompileOutput, ExitError>> {
+		None
+	}
+
+	fn addresses(&self) -> &[H160] {
+		&self.addresses
+	}
+}
 
 /// Stack-based executor.
-pub struct StackExecutor<'config, S> {
+pub struct StackExecutor<'config, S, P: Precompiles> {
```

Comment:
@sorpaas I guess that I misunderstood. So, if I change it so that every time we use `call` and `create`, I could pass in the precompile set there but, that may be a bit cumbersome.

Another option would be to make a struct called `Precompiles`, over a trait, which you then push every single `fn` and other data...?

The issue is for one of the EIPs for Berlin, we need to know the addresses for the precompiles so this change is absolutely necessary.

Reply:
That should be fine. I haven't found which EIP you're referring to but we can just add the parameter to the precompile function if needed.

What I really don't want to happen, however, is that the precompile type being dependent on executor type, like having the executor state type passed to precompile. In that case we get circle type dependency.

Reply:
I think you can just change the precompile interface to the following to avoid the generic.

```
pub type Precompile =
	(H160, fn(&[u8], Option<u64>, &Context) -> Option<Result<PrecompileOutput, ExitError>>);
```

Or better -- make it a struct to be more descriptive.

Reply:
This should do it: https://github.com/rust-blockchain/evm/pull/40/commits/5bad029315c35c77841b7a457c2a59ddcf370db2